src/lib/optimizer/strategy/stored_table_column_alignment_rule.hpp

```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+#include <unordered_map>
+#include <vector>
+
+#include "abstract_rule.hpp"
+#include "logical_query_plan/stored_table_node.hpp"
+
+namespace opossum {
+
+// Modified hash code generation for StoredTableNodes where column pruning information is omitted. Struct is used to
+// enable hash based containers containing std::shared_ptr<StoredTableNode>.
+struct StoredTableNodeSharedPtrHash final {
+  size_t operator()(const std::shared_ptr<StoredTableNode>& node) const {
+    size_t hash{0};
+    boost::hash_combine(hash, node->table_name);
+    for (const auto& pruned_chunk_id : node->pruned_chunk_ids()) {
+      boost::hash_combine(hash, static_cast<size_t>(pruned_chunk_id));
+    }
+    return hash;
+  }
+};
+
+// Modified equals evaluation code for StoredTableNodes where column pruning information is omitted. Struct is used to
+// enable hash based containers containing std::shared_ptr<StoredTableNode>.
+struct StoredTableNodeSharedPtrEqual final {
+  size_t operator()(const std::shared_ptr<StoredTableNode>& lhs, const std::shared_ptr<StoredTableNode>& rhs) const {
+    return lhs == rhs || (lhs->table_name == rhs->table_name && lhs->pruned_chunk_ids() == rhs->pruned_chunk_ids());
+  }
+};
+
+template <typename Value>
+using StoredTableNodeUnorderedMap = std::unordered_map<std::shared_ptr<StoredTableNode>, Value,
+                                                       StoredTableNodeSharedPtrHash, StoredTableNodeSharedPtrEqual>;
+
+// The PQP sub plan memoization for StoredTableNodes is sensitive to the node's table name, set of pruned chunks and
+// set of pruned columns. Consequently, for multiple nodes with the same table name, chunk and column pruning, only one
+// GetTable operator is created and executed. In some queries, the ColumnPruningRule and ChunkPruningRule provide an
+// LQP with multiple StoredTableNodes where the table names and sets of pruned chunks are equal but the sets of pruned
+// columns are different. This leads to the creation and execution of different GetTable operators.
+//
+// This rule identifies StoredTableNodes with equal table names and sets of pruned chunks and aligns the pruned columns
+// so that the sets of pruned columns are equal which means that the PQP sub plan memoization is effective for theses
+// StoredTableNodes and only one GetTable operator is created.
```

Comment:
Suggestion, but i am not yet sure about it: This rule identifies StoredTableNodes that differ only in the pruned columns. It then intersects the lists of pruned columns. While this means that some columns are left unpruned, it makes the job easier for the memoization in the LQPTranslator. In our experiments, this has led to significant performance improvements and negligible reductions.

Reply:
I'd agree with that statement.